From 59cf7857d08cf111ce0575fff7d2c11351607358 Mon Sep 17 00:00:00 2001
From: Bernhard Kaindl <bernhard.kaindl@cloud.com>
Date: Tue, 26 Aug 2025 12:00:00 +0200
Subject: [PATCH v2 4/7] xen/page_alloc: Pass node to adjust_tot_pages,
 set_outstanding_pages

domain_adjust_tot_pages() was used to adjust the total pages of a domain
and when adding pages for updating outstanding claims, the NUMA node
needs to be passed. For decreasing the claims, a NUMA node is not used.

To avoid passing NUMA_NO_NODE to each call that reduces the tot_pages,
switch to static inline domain_decrease_total_pages() for decreasing
tot_pages. This improves the legibility of the calls and avoids raising
eyebrows for lots of calls where NUMA_NO_NODE would be passed to a call
to update the tot_pages. With the comment added to the static inline
the situation is explained for all reviewers and and the legibility
of the code is improved as a bonus.

The old function name domain_adjust_tot_pages() is removed and the new
function is renamed to domain_update_total_pages(). It adds the node
argument and is only used for call sites that increase the total pages,
which need to pass the NUMA node of the pages added to the domain.

For consistency, also add domain_increase_total_pages() as a static inline
for calls that increase the tot_pages, which also need to pass the NUMA node.

domain_set_outstanding_pages() sets the amount of outstanding claims
of a domain. We pass the node on which to to stake an claim,
or NUMA_NO_NODE for host-wide claims.

No functional change, as neither function uses the arguments
for anything yet. It's a prerequisite to simplify for the next
patch that introduces per-node claim counts.

Signed-off-by: Bernhard Kaindl <bernhard.kaindl@cloud.com>

diff --git a/xen/arch/x86/mm.c b/xen/arch/x86/mm.c
index beffccf1a3b4..3db00fb4d443 100644
--- a/xen/arch/x86/mm.c
+++ b/xen/arch/x86/mm.c
@@ -4452,7 +4452,7 @@ int steal_page(
     page_list_del(page, &d->page_list);
 
     /* Unlink from original owner. */
-    if ( !(memflags & MEMF_no_refcount) && !domain_adjust_tot_pages(d, -1) )
+    if ( !(memflags & MEMF_no_refcount) && !domain_decrease_total_pages(d, 1) )
         drop_dom_ref = true;
 
     nrspin_unlock(&d->page_alloc_lock);
diff --git a/xen/arch/x86/mm/mem_sharing.c b/xen/arch/x86/mm/mem_sharing.c
index da28266ef076..3086cf7e2433 100644
--- a/xen/arch/x86/mm/mem_sharing.c
+++ b/xen/arch/x86/mm/mem_sharing.c
@@ -720,7 +720,7 @@ static int page_make_sharable(struct domain *d,
     if ( !validate_only )
     {
         page_set_owner(page, dom_cow);
-        drop_dom_ref = !domain_adjust_tot_pages(d, -1);
+        drop_dom_ref = !domain_decrease_total_pages(d, 1);
         page_list_del(page, &d->page_list);
     }
 
@@ -766,7 +766,7 @@ static int page_make_private(struct domain *d, struct page_info *page)
     ASSERT(page_get_owner(page) == dom_cow);
     page_set_owner(page, d);
 
-    if ( domain_adjust_tot_pages(d, 1) == 1 )
+    if ( domain_increase_total_pages(d, page_to_nid(page), 1) == 1 )
         get_knownalive_domain(d);
     page_list_add_tail(page, &d->page_list);
     nrspin_unlock(&d->page_alloc_lock);
diff --git a/xen/common/domain.c b/xen/common/domain.c
index 150eea904b1b..eb0ad8da8323 100644
--- a/xen/common/domain.c
+++ b/xen/common/domain.c
@@ -1083,7 +1083,7 @@ int domain_kill(struct domain *d)
         rspin_barrier(&d->domain_lock);
         argo_destroy(d);
         vnuma_destroy(d->vnuma);
-        domain_set_outstanding_pages(d, 0);
+        domain_set_outstanding_pages(d, NUMA_NO_NODE, 0);
         /* fallthrough */
     case DOMDYING_dying:
         rc = domain_teardown(d);
diff --git a/xen/common/grant_table.c b/xen/common/grant_table.c
index a8fc3b7aef43..17962c70d13a 100644
--- a/xen/common/grant_table.c
+++ b/xen/common/grant_table.c
@@ -2429,7 +2429,7 @@ gnttab_transfer(
         }
 
         /* Okay, add the page to 'e'. */
-        if ( unlikely(domain_adjust_tot_pages(e, 1) == 1) )
+        if ( unlikely(domain_increase_total_pages(e, page_to_nid(page), 1) == 1) )
             get_knownalive_domain(e);
 
         /*
@@ -2455,7 +2455,7 @@ gnttab_transfer(
              * page in the page total
              */
             nrspin_lock(&e->page_alloc_lock);
-            drop_dom_ref = !domain_adjust_tot_pages(e, -1);
+            drop_dom_ref = !domain_decrease_total_pages(e, 1);
             nrspin_unlock(&e->page_alloc_lock);
 
             if ( okay /* i.e. e->is_dying due to the surrounding if() */ )
diff --git a/xen/common/memory.c b/xen/common/memory.c
index 73469dce308c..287d383f3cdb 100644
--- a/xen/common/memory.c
+++ b/xen/common/memory.c
@@ -773,7 +773,7 @@ static long memory_exchange(XEN_GUEST_HANDLE_PARAM(xen_memory_exchange_t) arg)
 
                 nrspin_lock(&d->page_alloc_lock);
                 drop_dom_ref = (dec_count &&
-                                !domain_adjust_tot_pages(d, -dec_count));
+                                !domain_decrease_total_pages(d, dec_count));
                 nrspin_unlock(&d->page_alloc_lock);
 
                 if ( drop_dom_ref )
@@ -1683,7 +1683,8 @@ long do_memory_op(unsigned long cmd, XEN_GUEST_HANDLE_PARAM(void) arg)
             rc = -EINVAL;
 
         if ( !rc )
-            rc = domain_set_outstanding_pages(d, reservation.nr_extents);
+            rc = domain_set_outstanding_pages(d, NUMA_NO_NODE,
+                                              reservation.nr_extents);
 
         rcu_unlock_domain(d);
 
diff --git a/xen/common/page_alloc.c b/xen/common/page_alloc.c
index 644f705abb04..cd90b2f7f485 100644
--- a/xen/common/page_alloc.c
+++ b/xen/common/page_alloc.c
@@ -496,7 +496,8 @@ static long outstanding_claims; /* total outstanding claims by all domains */
  * - When pages were freed, we do not increase outstanding claims.
  * - On a domain's claims update, global outstanding_claims are updated as well.
  */
-unsigned long domain_adjust_tot_pages(struct domain *d, long pages)
+unsigned long domain_update_total_pages(
+    struct domain *d, nodeid_t node, long pages)
 {
     unsigned long adjustment;
 
@@ -529,7 +530,8 @@ out:
     return d->tot_pages;
 }
 
-int domain_set_outstanding_pages(struct domain *d, unsigned long pages)
+int domain_set_outstanding_pages(
+    struct domain *d, nodeid_t node, unsigned long pages)
 {
     int ret = -ENOMEM;
     unsigned long claim, avail_pages;
@@ -2621,6 +2623,8 @@ int assign_pages(
 
     if ( !(memflags & MEMF_no_refcount) )
     {
+        nodeid_t node = page_to_nid(&pg[0]);
+
         if ( unlikely(d->tot_pages + nr < nr) )
         {
             gprintk(XENLOG_INFO,
@@ -2632,7 +2636,9 @@ int assign_pages(
             goto out;
         }
 
-        if ( unlikely(domain_adjust_tot_pages(d, nr) == nr) )
+        ASSERT(node == page_to_nid(&pg[nr - 1]));
+
+        if ( unlikely(domain_increase_total_pages(d, node, nr) == nr) )
             get_knownalive_domain(d);
     }
 
@@ -2765,7 +2771,7 @@ void free_domheap_pages(struct page_info *pg, unsigned int order)
                 }
             }
 
-            drop_dom_ref = !domain_adjust_tot_pages(d, -(1 << order));
+            drop_dom_ref = !domain_decrease_total_pages(d, 1 << order);
 
             rspin_unlock(&d->page_alloc_lock);
 
@@ -2978,7 +2984,7 @@ void free_domstatic_page(struct page_info *page)
 
     arch_free_heap_page(d, page);
 
-    drop_dom_ref = !domain_adjust_tot_pages(d, -1);
+    drop_dom_ref = !domain_decrease_total_pages(d, 1);
 
     unprepare_staticmem_pages(page, 1, scrub_debug);
 
diff --git a/xen/include/xen/mm.h b/xen/include/xen/mm.h
index 16f733281af3..65d02f8945fa 100644
--- a/xen/include/xen/mm.h
+++ b/xen/include/xen/mm.h
@@ -66,6 +66,7 @@
 #include <xen/mm-frame.h>
 #include <xen/types.h>
 #include <xen/list.h>
+#include <xen/numa.h>
 #include <xen/spinlock.h>
 #include <xen/perfc.h>
 #include <public/memory.h>
@@ -127,10 +128,39 @@ mfn_t xen_map_to_mfn(unsigned long va);
  * page range in Xen virtual address space.
  */
 int populate_pt_range(unsigned long virt, unsigned long nr_mfns);
-/* Claim handling */
-unsigned long __must_check domain_adjust_tot_pages(struct domain *d,
-    long pages);
-int domain_set_outstanding_pages(struct domain *d, unsigned long pages);
+
+/* Adjust the domain size. When adding pages, also update remaining claims */
+unsigned long __must_check domain_update_total_pages(struct domain *d,
+    nodeid_t node, long pages);
+
+/*
+ * When pages are freed: Decrease the page count of the domain.
+ *
+ * Adding pages is handled by domain_update_total_pages() instead:
+ * It also needs the NUMA node to update the outstanding NUMA node claims.
+ *
+ * This is something that is not needed when pages are freed, because
+ * the outstanding claims are not incremented when pages are freed.
+ *
+ * With this inline, we keep the callers that free pages functionally
+ * unchanged, and don't touch them as they do not need to pass a node.
+ */
+static inline unsigned long __must_check
+domain_decrease_total_pages(struct domain *d, unsigned long pages)
+{
+    /* When freeing pages, we don't increase claims and can pass NUMA_NO_NODE */
+    return domain_update_total_pages(d, NUMA_NO_NODE, -pages);
+}
+
+static inline unsigned long __must_check
+domain_increase_total_pages(struct domain *d, nodeid_t node, unsigned long pages)
+{
+    /* When pages are added, pass the NUMA node as we need to update claims */
+    return domain_update_total_pages(d, node, pages);
+}
+
+int domain_set_outstanding_pages(struct domain *d, nodeid_t node,
+                                 unsigned long pages);
 void get_outstanding_claims(uint64_t *free_pages, uint64_t *outstanding_pages);
 
 /* Domain suballocator. These functions are *not* interrupt-safe.*/
diff --git a/xen/include/xen/sched.h b/xen/include/xen/sched.h
index 882db910f77d..a1321a108db5 100644
--- a/xen/include/xen/sched.h
+++ b/xen/include/xen/sched.h
@@ -402,7 +402,7 @@ struct domain
 #endif
 
     /*
-     * This field should only be directly accessed by domain_adjust_tot_pages()
+     * This field should only be directly accessed by domain_update_total_pages()
      * and the domain_tot_pages() helper function defined below.
      */
     unsigned int     tot_pages;
